backend/amd64/cfg_stack_operands.ml

```diff
@@ -0,0 +1,218 @@
+
+let may_use_stack_operand_for_second_argument
+  : type a . Reg.t Reg.Tbl.t -> a Cfg.instruction -> stack_operands_rewrite
+  = fun map instr ->
+  if debug then begin
+    check_lengths instr ~of_arg:2 ~of_res:1;
+    check_same "res(0)" instr.res.(0) "arg(0)" instr.arg.(0);
+  end;
+  begin match is_spilled instr.arg.(1) with
+  | false -> ()
+  | true ->
+    use_stack_operand map instr.arg 1;
+  end;
+  May_still_have_spilled_registers
+
+let may_use_stack_operand_for_only_argument
+  : type a . has_result:bool -> Reg.t Reg.Tbl.t -> a Cfg.instruction -> stack_operands_rewrite
+  = fun ~has_result map instr ->
+  if debug then check_lengths instr ~of_arg:1 ~of_res:(if has_result then 1 else 0);
+  begin match is_spilled instr.arg.(0) with
+  | false -> ()
+  | true ->
+    use_stack_operand map instr.arg 0
+  end;
+  if has_result then
+    May_still_have_spilled_registers
+  else
+    All_spilled_registers_rewritten
+
+let may_use_stack_operand_for_only_result
+  : type a . Reg.t Reg.Tbl.t -> a Cfg.instruction -> stack_operands_rewrite
+  = fun map instr ->
+  if debug then check_lengths instr ~of_arg:0 ~of_res:1;
+  begin match is_spilled instr.res.(0) with
+  | false ->
+    All_spilled_registers_rewritten
+  | true ->
+    use_stack_operand map instr.res 0;
+    All_spilled_registers_rewritten
+  end
+
+let may_use_stack_operand_for_result
+  : type a . num_args:int -> Reg.t Reg.Tbl.t -> a Cfg.instruction -> stack_operands_rewrite
+  = fun ~num_args map instr ->
+  if debug then begin
+    check_lengths instr ~of_arg:num_args ~of_res:1;
+  end;
+  begin match is_spilled instr.res.(0) with
+  | false -> ()
+  | true ->
+    if Reg.same instr.arg.(0) instr.res.(0) then begin
+      use_stack_operand map instr.arg 0;
+    end;
+    use_stack_operand map instr.res 0;
+  end;
+  May_still_have_spilled_registers
+
+type result =
+  | No_result
+  | Result_can_be_on_stack
+  | Result_cannot_be_on_stack
+
+let binary_operation
+  : type a . Reg.t Reg.Tbl.t -> a Cfg.instruction -> result -> stack_operands_rewrite
+  = fun map instr result ->
+  if debug then begin
+    match result with
+    | No_result ->
+      check_lengths instr ~of_arg:2 ~of_res:0
+    | Result_can_be_on_stack ->
+      check_lengths instr ~of_arg:2 ~of_res:1;
+      check_same "res(0)" instr.res.(0) "arg(0)" instr.arg.(0)
+    | Result_cannot_be_on_stack ->
+      check_lengths instr ~of_arg:2 ~of_res:1
+  end;
+  begin match is_spilled instr.arg.(0), is_spilled instr.arg.(1) with
+  | false, false ->
+    begin match result with
+    | No_result | Result_can_be_on_stack ->
+      All_spilled_registers_rewritten
+    | Result_cannot_be_on_stack ->
+      May_still_have_spilled_registers
+    end
+  | false, true ->
+    use_stack_operand map instr.arg 1;
+    begin match result with
+    | No_result ->
+      All_spilled_registers_rewritten
+    | Result_can_be_on_stack | Result_cannot_be_on_stack ->
+      May_still_have_spilled_registers
+    end
+  | true, false ->
```

Comment:
I think we need a comment here as to why this case is not symmetric to the above. I think it has to do with the fact that we can use both arg.(0) and res.(0) on stack when  they are in the same  location, but it is not obvious here, it's implicit from `Selection` which assigns the same Reg.t to both places.

Reply:
Added a short comment.

Reply:
Would it be better to add a check that arg.(0) and res.(0) are the same in the `Result_can_be_on_stack` case, and only use stack locations for them under this condition?

Reply:
I am not sure what you mean, there is such a check on line 77
as far as I understand.

Reply:
sorry, I missed it (because it was under debug).